Draft: Modular block request handler

rahulksnv opened the pull request:
If a substrate user wants to implement a custom block relay protocol (like Compact Blocks, Graphene, etc), it is not currently possible. This is because the block request handler is tightly coupled with the rest of the stack. This change makes the block request handler modular.

1. `ChainSync` currently calls into the block request handler directly. Instead, move the block request handler behind a trait. This allows new protocols to be plugged into `ChainSync`.
2. `BuildNetworkParams` is changed so that custom relay protocol implementations can be (optionally) passed in during network creation time. If custom protocol is not specified, it defaults to the existing block handler
3. `BlockServer` and `BlockDownloader` traits are introduced for the protocol implementation. The existing block handler has been changed to implement these traits

NOTE: 
1. This is the change we are submitting against upstream next. This PR is only to get feedback for now
2.  This is a breaking change for the subspace repo, as corresponding changes are needed to make it compile. That will be submitted as a separate PR

nazar-pc commented on client/transaction-pool/api/src/lib.rs:
```diff
@@ -186,7 +187,7 @@ pub trait TransactionPool: Send + Sync {
 	type Block: BlockT;
 	/// Transaction hash type.
-	type Hash: Hash + Eq + Member + Serialize + DeserializeOwned;
+	type Hash: Hash + Eq + Member + Serialize + DeserializeOwned + Codec;
```
Why is `Codec` necessary here?

rahulksnv replied:
This is needed so that the TxnHash<> can be serialized and sent in the compact block response. Technically this is not part of the modular changes, I will call this out in the upstream patch

nazar-pc replied:
I don't see any changes with `TxnHash` in this PR, so I'm a bit confused. Is it necessary downstream in Subspace? If so, can't we add corresponding bound there instead?

rahulksnv replied:
Hm, this is used in the compact block implementation (you won't see any changes here as expected). Just like `Block/Hash/Extrinsic` in https://paritytech.github.io/substrate/master/sp_runtime/traits/trait.Block.html have `Codec`, and  we do `encode()/decode()` on these, we do the same on TxHash as well. Does it answer your question?

nazar-pc replied:
Not a huge issue of course, I just found it weird that after refactoring we suddenly need a bound that we didn't need before. If it is used in compact blocks (hence not necessary for this PR specifically), then maybe there is a way to specify it in compact blocks only.

rahulksnv replied:
This is actually cleaner(makes the intent clear that this can be serialized), and reusable and others as well. Also avoids another upstream PR cycle later if needed. I am going through it now, lot of steps/time consuming

nazar-pc reviewed (commented):
Commits should be squashed, easier to review that way. Also unrelated `Cargo.lock` changes should be reverted. Otherwise makes sense.

rahulksnv commented:
> Commits should be squashed, easier to review that way. Also unrelated `Cargo.lock` changes should be reverted. Otherwise makes sense.

 The commit history you see now is after squashing, those are two logical steps in the process and needed. And I deleted the Cargo.lock and regenerated, this is what it generates even though I have a single change in Cargo.toml

nazar-pc reviewed (commented):
> The commit history you see now is after squashing, those are two logical steps in the process and needed.

Second step is "Make protobufs private", which would have not been necessary in the first place if first step didn't make those protobufs public, so by squashing those together you should get less changes in total. I either miss something or those two commits are essentially the same step.

> And I deleted the Cargo.lock and regenerated

That is the wrong way to do it though. `Cargo.lock` was already in the repo. Instead of deleting it (which will cause new `Cargo.lock` being created with latest versions of packages in it) you need to revert to the version that was before your changes and run `cargo c` or whatever to get `Cargo.lock` **updated**, then it will only have minimal changes you actually want.

rahulksnv commented:
> > And I deleted the Cargo.lock and regenerated
> 
> That is the wrong way to do it though. `Cargo.lock` was already in the repo. Instead of deleting it (which will cause new `Cargo.lock` being created with latest versions of packages in it) you need to revert to the version that was before your changes and run `cargo c` or whatever to get `Cargo.lock` **updated**, then it will only have minimal changes you actually want.
Thanks, cleaned it up
